torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3137,6 +3111,30 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 # github.com/NVIDIA/apex/blob/master/apex/parallel/distributed.py
                 orig_grad_data.record_stream(self._streams["post_backward"])
 
+    def _cast_grad_to_full_param_dtype(
+        self,
+        grad: torch.Tensor,
+        full_param_dtype: torch.dtype,
+    ):
+        """
+        Casts gradients back to the full parameter dtype so that the optimizer
+        step runs with that dtype. This performs an actual cast if
+        1. parameters were in reduced precision during the forward since then
+        gradients would be in that reduced precision, or
+        2. parameters were not in reduced precision but gradients were in
+        reduced precision for communication.
+        """
+        self._assert_state(TrainingState_.BACKWARD_POST)
+        if (
+            self._mixed_precision_enabled_for_params()
+            or self._mixed_precision_enabled_for_reduce()
+        ):
+            low_prec_grad_data = grad.data
+            grad.data = grad.data.to(dtype=full_param_dtype)
```

Comment:
`non_blocking=True`?

Reply:
`non_blocking=True` seems safe here, but I am also not entirely sure. The [discussion online](https://discuss.pytorch.org/t/should-we-set-non-blocking-to-true/38234/17?u=agu) mostly refers to CPU <-> GPU transfers, not GPU <-> GPU transfers with different `dtype`. I am going to leave it as `non_blocking=False` (default) since I am not fully confident it will be safe.